app/assets/javascripts/listings/ListingDataService.js.coffee

```diff
@@ -0,0 +1,434 @@
+
+  Service.cleanListings = (listings) ->
+    _.map listings, (listing) ->
+      # fallback for fixing the layout when a listing is missing an image
+      listing.imageURL ?= 'https://unsplash.it/g/780/438'
+    _.filter listings, (listing) ->
+      !_.includes(MAINTENANCE_LISTINGS, listing.Id)
+
+  Service.groupListings = (listings) ->
+    openListings = []
+    openMatchListings = []
+    openNotMatchListings = []
+    closedListings = []
+    lotteryResultsListings = []
+
+    listings.forEach (listing) ->
+      if ListingIdentityService.isOpen(listing)
+        # All Open Listings Array
+        openListings.push(listing)
+        if listing.Does_Match
+          openMatchListings.push(listing)
+        else
+          openNotMatchListings.push(listing)
+      else
+        if ListingLotteryService.lotteryIsUpcoming(listing)
+          closedListings.push(listing)
+        else
+          lotteryResultsListings.push(listing)
+
+    angular.copy(Service.sortListings(openListings, 'openListings'), Service.openListings)
+    angular.copy(Service.sortListings(openMatchListings, 'openMatchListings'), Service.openMatchListings)
+    angular.copy(Service.sortListings(openNotMatchListings, 'openNotMatchListings'), Service.openNotMatchListings)
+    angular.copy(Service.sortListings(closedListings, 'closedListings'), Service.closedListings)
+    angular.copy(Service.sortListings(lotteryResultsListings, 'lotteryResultsListings'), Service.lotteryResultsListings)
+
+  Service.sortListings = (listings, type) ->
+    # openListing types
+    if ['openListings', 'openMatchListings', 'openNotMatchListings'].indexOf(type) > -1
+      _.sortBy listings, (i) -> moment(i.Application_Due_Date)
+    # closedListing types
+    else if ['closedListings', 'lotteryResultsListings'].indexOf(type) > -1
+      listings = _.sortBy listings, (i) ->
+        # fallback to Application_Due_Date, really only for the special case of First Come First Serve
+        moment(i.Lottery_Results_Date || i.Application_Due_Date)
+      # lotteryResults get reversed (latest lottery results date first)
+      if type == 'lotteryResultsListings' then _.reverse listings else listings
+
+  Service.getListingsByIds = (ids, checkFavorites = false) ->
+    angular.copy([], Service.listings)
+    params = {params: {ids: ids.join(',') }}
+    $http.get("/api/v1/listings.json", params).success((data, status, headers, config) ->
+      listings = if data and data.listings then data.listings else []
+      angular.copy(listings, Service.listings)
+      Service.checkFavorites() if checkFavorites
+    ).error( (data, status, headers, config) ->
+      return
+    )
+
+  Service.isAcceptingOnlineApplications = (listing) ->
+    return false if _.isEmpty(listing)
+    return false if ListingLotteryService.lotteryComplete(listing)
+    return false unless ListingIdentityService.isOpen(listing)
+    return listing.Accepting_Online_Applications
+
+  Service.getListingAndCheckIfOpen = (id) ->
```

Comment:
this seems oddly specific and the name doesn't really communicate that if there's no listing or if it's not open it's going to change the state. It also seems to only be used in angularRoutes. Would it make more sense to move these if/else statements directly into angular routes?

Reply:
i would be fine with moving it to angularRoutes, would you like to do that?